c2rust-analyze/src/rewrite/apply.rs

```diff
@@ -0,0 +1,506 @@
+                self.emit_str(">");
+            }
+        }
+    }
+
+    fn emit_bytes(&mut self, lo: BytePos, hi: BytePos) {
+        assert!(
+            self.file.start_pos <= lo && hi <= self.file.end_pos,
+            "bytes {:?} .. {:?} are out of range for file {:?}",
+            lo,
+            hi,
+            self.file.name
+        );
+        let src = self
+            .file
+            .src
+            .as_ref()
+            .unwrap_or_else(|| panic!("source is not available for file {:?}", self.file.name));
+        self.emit_str(&src[lo.0 as usize..hi.0 as usize]);
+    }
+
+    fn emit_span_with_rewrites(&mut self, span: Span, rts: &[RewriteTree]) {
+        let (_, overlap, _) = partition_nodes(rts, span);
+
+        let mut pos = span.lo();
+        for rt in overlap {
+            // Every child node is contained by the span of some `Rewrite::Identity` or
+            // `Rewrite::Sub` in its parent node.
+            debug_assert!(span.contains(rt.span));
+
+            self.emit_bytes(pos, rt.span.lo());
+            self.emit_rewrite(
+                &rt.rw,
+                0,
+                &mut |slf| slf.emit_span_with_rewrites(rt.span, &rt.children),
+                &mut |slf, subexpr_span| slf.emit_span_with_rewrites(subexpr_span, &rt.children),
+            );
+            pos = rt.span.hi();
+        }
+
+        self.emit_bytes(pos, span.hi());
+    }
+}
+
+/// Apply rewrites `rws` to the source files covered by their `Span`s.  Returns a map giving the
+/// rewritten source code for each file that contains at least one rewritten `Span`.
+pub fn apply_rewrites(
+    source_map: &SourceMap,
+    rws: Vec<(Span, Rewrite)>,
+) -> HashMap<FileName, String> {
+    let (rts, errs) = RewriteTree::build(rws);
+    for (span, rw, err) in errs {
+        eprintln!(
+            "{:?}: warning: failed to apply rewrite {:?}: {:?}",
+            span, rw, err
+        );
+    }
+
+    let mut new_src = HashMap::new();
+    let mut rts = &rts as &[RewriteTree<Span>];
+    while rts.len() > 0 {
+        let file = source_map.lookup_source_file(rts[0].span.lo());
+        let idx = rts
+            .iter()
+            .position(|rt| rt.span.lo() >= file.end_pos)
+            .unwrap_or(rts.len());
+        assert!(idx > 0);
+        let (file_rts, rest) = rts.split_at(idx);
+        rts = rest;
+
+        let mut buf = String::new();
+        let mut emit = |s: &str| {
+            buf.push_str(s);
+        };
+        let mut emitter = Emitter {
+            file: &file,
+            emit: &mut emit,
+        };
+        let file_span = Span::new(file.start_pos, file.end_pos, SyntaxContext::root(), None);
+        emitter.emit_span_with_rewrites(file_span, file_rts);
+
+        new_src.insert(file.name.clone(), buf);
+    }
+
+    new_src
+}
+
+#[cfg(test)]
+mod test {
+    use super::*;
+
+    type FakeSpan = (u32, u32);
+
+    fn mk(start: u32, end: u32, i: usize) -> (FakeSpan, Rewrite<FakeSpan>) {
```

Comment:
are `start` and `end` equivalent to `lo` and `hi`? if so, i think it'd be clearer to having naming consistency